src/workerd/api/streams/standard.c++

```diff
@@ -3022,15 +3078,30 @@ private:
 
           if constexpr (kj::isSameType<T, ByteReadable>()) {
             jsg::BackingStore backing = bufferSource.detach(js);
-            return backing.asArrayPtr().attach(kj::mv(backing));
+            accumulator.add(backing.asArrayPtr().attach(kj::mv(backing)));
           } else if constexpr (kj::isSameType<T, ValueReadable>()) {
             // We do not detach in this case because, as bad as an idea as it is,
             // the stream spec does allow a single typedarray/arraybuffer instance
             // to be queued multiple times when using value-oriented streams.
-            return bufferSource.asArrayPtr().attach(kj::mv(bufferSource));
+            accumulator.add(bufferSource.asArrayPtr().attach(kj::mv(bufferSource)));
+          } else {
+            KJ_FAIL_ASSERT("incorrect readable type. how did that happen?");
           }
 
-          KJ_UNREACHABLE;
+          // We accumulate the bytes we've read in the accumulator. Once we've read
+          // at least Accumulator::THRESHOLD, or have Acculator::MAX_COUNT buffers
+          // collected, we'll release the collected bytes to the destination, otherwise
+          // we'll skip to another read to pull more bytes.
+          if (accumulator.total >= Accumulator::THRESHOLD ||
```

Comment:
Optional: the `Accumulator` class could encapsulate this logic for a bit less repetition in documentation and one point of policy, e.g. have an`Accumulator::tryRelease()` that conditionally returns the accumulated buffers ather than `Accumulator::release()`.

Reply:
I guess tryRelease would be gated on `pendingClose` in addition to bytes or buffer counts.